easybuild/easyblocks/generic/cargo.py

```diff
@@ -0,0 +1,136 @@
+            'offline': [True, "Build offline", CUSTOM],
+            'lto': [False, "Build with link time optimization", CUSTOM],
+            'crates': [[], "List of (crate, version) tuples to use", CUSTOM],
+        })
+
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Constructor for Cargo easyblock."""
+        super(Cargo, self).__init__(*args, **kwargs)
+        env.setvar('CARGO_HOME', os.path.join(self.builddir, '.cargo'))
+        env.setvar('RUSTC', 'rustc')
+        env.setvar('RUSTDOC', 'rustdoc')
+        env.setvar('RUSTFMT', 'rustfmt')
+        optarch = build_option('optarch')
+        if not optarch:
+            optarch = 'native'
+        env.setvar('RUSTFLAGS', '-C target-cpu=%s' % optarch)
+        env.setvar('RUST_LOG', 'DEBUG')
+        env.setvar('RUST_BACKTRACE', '1')
+
+        # Populate sources from "crates" list of tuples
+        sources = self.cfg['sources']
+        for crate, version in self.cfg['crates']:
+            sources.append({
+                'download_filename': crate + '/' + version + '/download',
+                'filename': crate + '-' + version + '.tar.gz',
+                'source_urls': [CRATESIO_SOURCE],
+                'alt_location': 'crates.io',
+            })
+        self.cfg.update('sources', sources)
+
+    def configure_step(self):
+        pass
+
+    def extract_step(self):
+        """Populate all vendored deps with required .cargo-checksum.json"""
+        EasyBlock.extract_step(self)
+        for source in self.src:
+            dirname = source["name"].rsplit('.', maxsplit=2)[0]
+            self.log.info('creating .cargo-checksums.json file for : %s', dirname)
+            chksum = compute_checksum(source['path'], checksum_type='sha256')
+            chkfile = '%s/%s/.cargo-checksum.json' % (self.builddir, dirname)
+            write_file(chkfile, '{"files":{},"package":"%s"}' % chksum)
+
+    @property
+    def profile(self):
+        return 'debug' if self.toolchain.options.get('debug', None) else 'release'
+
+    def build_step(self):
+        """Build with cargo"""
+        parallel = ''
+        if self.cfg['parallel']:
+            parallel = "-j %s" % self.cfg['parallel']
+
+        tests = ''
+        if self.cfg['enable_tests']:
+            parallel = "--tests"
+
+        offline = ''
+        if self.cfg['offline']:
+            parallel = "--offline"
+            # Replace crates-io with vendored sources
+            write_file('.cargo/config.toml', '[source.crates-io]\ndirectory=".."', append=True)
+
+        lto = ''
+        if self.cfg['lto']:
+            parallel = '--config profile.%s.lto=true' % self.profile
+
+        run_cmd('rustc --print cfg', log_all=True, simple=True)  # for tracking in log file
+        cmd = '%s cargo build --profile=%s %s %s %s %s %s' % (
+            self.cfg['prebuildopts'], self.profile, offline, lto, tests, parallel, self.cfg['buildopts'])
```

Comment:
this can be improved a bit (same below for test + install commands)

```suggestion
        cmd = ' '.join([
            self.cfg['prebuildopts'],
            "cargo build",
            "--profile=" + self.profile,
            offline,
            lto,
            tests,
            parallel,
            self.cfg['buildopts'],
        ])